cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -372,29 +376,32 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
 	namespaceKubeClient := client("namespace-controller")
 	namespaceClientPool := dynamic.NewClientPool(restclient.AddUserAgent(kubeconfig, "namespace-controller"), restMapper, dynamic.LegacyAPIPathResolverFunc)
 	// TODO: consider using a list-watch + cache here rather than polling
-	var gvrFn func() ([]schema.GroupVersionResource, error)
+	discoverResourcesFn := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources
 	rsrcs, err := namespaceKubeClient.Discovery().ServerResources()
 	if err != nil {
 		glog.Fatalf("Failed to get group version resources: %v", err)
 	}
+	tprFound := false
+searchThirdPartyResource:
 	for _, rsrcList := range rsrcs {
 		for ix := range rsrcList.APIResources {
 			rsrc := &rsrcList.APIResources[ix]
 			if rsrc.Kind == "ThirdPartyResource" {
-				gvrFn = namespaceKubeClient.Discovery().ServerPreferredNamespacedResources
+				tprFound = true
+				break searchThirdPartyResource
 			}
 		}
 	}
-	if gvrFn == nil {
-		gvr, err := namespaceKubeClient.Discovery().ServerPreferredNamespacedResources()
+	if !tprFound {
+		resourceLists, err := discoverResourcesFn()
 		if err != nil {
 			glog.Fatalf("Failed to get resources: %v", err)
 		}
-		gvrFn = func() ([]schema.GroupVersionResource, error) {
-			return gvr, nil
+		discoverResourcesFn = func() ([]*unversioned.APIResourceList, error) {
+			return resourceLists, nil
 		}
 	}
-	namespaceController := namespacecontroller.NewNamespaceController(namespaceKubeClient, namespaceClientPool, gvrFn, s.NamespaceSyncPeriod.Duration, v1.FinalizerKubernetes)
+	namespaceController := namespacecontroller.NewNamespaceController(namespaceKubeClient, namespaceClientPool, discoverResourcesFn, s.NamespaceSyncPeriod.Duration, v1.FinalizerKubernetes)
```

Comment:
blech.  If the namespace controller relies on discovery, it should have a discovery client.  Debt for later I suppose.